### PR TITLE
added basic support WithAnnotations

### PR DESCRIPTION
**Pull request by arekbal**
# Info
This Pull Request is related to issue no. #29 

Did #26 as well.

# Changes
- added WithAnnotations on Ensure
// might be invoked by default in the future for each Ensure

### CODE REVIEW

File: tests/Valit.Tests/UInt64/UInt64_Required_Tests.cs [hunk 1]
```diff
@@ -20,7 +19,7 @@ public void UInt64_Required_For_Nullable_Value_Throws_When_Null_Rule_Is_Given()
 
         [Theory]
         [InlineData(false, true)]
-        [InlineData(true, false)]
```
Comment (dbarwikowski) [inline tests/Valit.Tests/UInt64/UInt64_Required_Tests.cs:23]: Extra space

File: tests/Valit.Tests/UInt64/UInt64_IsLessThan_Tests.cs [hunk 2]
```diff
@@ -108,7 +108,7 @@ public void UInt64_IsLessThan_Returns_Proper_Results_For_Nullable_Value_And_Not_
         [InlineData(false, null, false)]     
         [InlineData(true, (ulong) 11, false)]     
         [InlineData(true, null, false)]     
-        public void UInt64_IsLessThan_Returns_Proper_Results_For_Nullable_Values(bool useNullValue, ulong? value,  bool expected)
```
Comment (dbarwikowski) [inline tests/Valit.Tests/UInt64/UInt64_IsLessThan_Tests.cs:111]: Extra space

File: src/Valit/Rules/ValitRules.cs [hunk 3]
```diff
@@ -2,93 +2,114 @@
...
-    public class ValitRules<TObject> : 
-        IValitRules<TObject>,
-        IValitRulesMessageProvider<TObject>, 
-        IValitRulesStrategyPicker<TObject> 
-        where TObject : class
+    public static class ValitRules
     {
-        private TObject _object;       
-        private readonly List<IValitRule<TObject>> _rules;
+        public static IValitRulesStrategyPicker<TTarget> CreateFor<TTarget>(TTarget @object) where TTarget : class
```
Comment (dbarwikowski) [inline src/Valit/Rules/ValitRules.cs:13]: Ideally, we would like to create stateless validator, which would be build once and run on many models. That's why we changed Create<TModel>(TModel model) to just Create()

File: src/Valit/Rules/IValitRules.cs [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
     {
-         IValitRules<TObject> Ensure<TProperty>(Func<TObject, TProperty> selector, Func<IValitRule<TObject, TProperty>,IValitRule<TObject, TProperty>> rule);
+         IValitRules<TObject> Ensure<TProperty>(Expression<Func<TObject, TProperty>> selector, Func<IValitRule<TObject, TProperty>,IValitRule<TObject, TProperty>> rule);
          IValitRules<TObject> For(TObject @object);
          IEnumerable<IValitRule<TObject>> GetTaggedRules();
          IEnumerable<IValitRule<TObject>> GetUntaggedRules();         
          IValitResult Validate();
          IValitResult Validate(params string[] tags);
-         IValitResult Validate(Func<IValitRule<TObject>, bool> predicate);
+         IValitResult Validate(Func<IValitRule<TObject>, bool> predicate);
```
Comment (dbarwikowski) [inline src/Valit/Rules/IValitRules.cs:15]: Not very usefull :)

File: src/Valit/Exceptions/ValitExceptionExtensions.cs [hunk 5]
```diff
@@ -6,22 +6,26 @@ namespace Valit
 {
     public static class ValitExceptionExtensions
     {
-        internal static void ThrowIfNull<T>(this T @object)
+        internal static T ThrowIfNull<T>(this T @object)
```
Comment (dbarwikowski) [inline src/Valit/Exceptions/ValitExceptionExtensions.cs:9]: I like what you did here :)

File: src/Valit/Rules/Extensions/_Valit/ValitRulePropertyExtensions.cs [hunk 6]
```diff
@@ -83,7 +86,7 @@ public static class ValitRulePropertyExtensions
 
             previousRuleAccessor.AddTags(tags);            
             return new ValitRule<TObject, TProperty>(accessor.PreviousRule);
-        }
+        }
```
Comment (dbarwikowski) [inline src/Valit/Rules/Extensions/_Valit/ValitRulePropertyExtensions.cs:89]: Spaces

File: src/Valit/Rules/Extensions/_Valit/ValitRulePropertyExtensions.cs [hunk 7]
```diff
@@ -1,5 +1,8 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using Valit.Expressions;
```
Comment (dbarwikowski) [inline src/Valit/Rules/Extensions/_Valit/ValitRulePropertyExtensions.cs:5]: Do we need those 3? :)

File: src/Valit/Rules/ValitRules.cs (hunk 3 above)
Comment (arekbal) [reply src/Valit/Rules/ValitRules.cs:13]: That is true... but in the meantime... while we have this '.For'... method.

**Review by GooRiOn — COMMENTED** [review]
I'm not quite sure whether we should have both For and CreateFor. As fa as I understand the second one is need to properly use ExpressionVisitor, but it could be a little bit confusing for folks. Maybe we should consider ValitRules and AnnotationValitRules?

**Comment by dbarwikowski** [comment]
Isn't this Pull request related to issue #26?

**Pull request closed without merging**